docs/docs/remark-ast.md

```diff
@@ -0,0 +1,270 @@
+
+```js:title=plugins/gatsby-remark-purple-headers/index.js
+module.exports = ({ markdownAST }, pluginOptions) => {
+  // Manipulate AST
+
+  return markdownAST
+}
+```
+
+The first parameter is all of the default properties that can be used in plugins (actions, store, getNodes, schema, etc.) plus a couple just for gatsby-transformer-remark plugins. The one to be focused on is the `markdownAST` field which is destructured in the code snippet above.
+
+As with other Gatsby plugins, the 2nd parameter is the pluginOptions which is obtained from the definition in `gatbsy-config.js` file.
+
+Finally, the function will return the markdownAST after the fields you wish to be edited are transformed.
+
+## Adding the plugin to your site
+
+You likely will want to grab `gatsby-source-filesystem` to grab the file nodes initially. in this example I am going to assume the markdown files we are working with exist in a `src/data/` directory.
+
+The plugin is now initially set so we can add it as a sub-plugin inside `gatsby-transformer-remark`
+
+```js:title=gatsby-config.js
+module.exports = {
+  plugins: [
+    {
+      resolve: "gatsby-source-filesystem",
+      options: {
+        name: `data`,
+        path: `${__dirname}/src/data/`,
+      },
+    },
+    {
+      resolve: `gatsby-transformer-remark`,
+      options: {
+        plugins: [`gatsby-remark-purple-headers`],
+      },
+    },
+  ],
+}
+```
+
+If we wanted to add some options, we could switch to the object syntax:
+
+```js
+{
+  resolve: `gatsby-remark-purple-headers`,
+  options: {
+    // Options here
+  }
+}
+```
+
+## Find and Modify Markdown Nodes
+
+When modifying nodes, you'll want to walk the tree and then implement new functionality on specific nodes.
+
+A node module to help with is [unist-util-visit](https://github.com/syntax-tree/unist-util-visit), a walker for `unist` nodes. For reference, Unist which stands for United Syntax Tree is a standard for markdown syntax trees and parsers that include well known parsers in the Gatsby world like Remark and MDX.
+
+As an example from `unist-util-visit`'s readme, it allows for an easy interface to visit particular nodes based on a particular type:
+
+```js
+var remark = require("remark")
+var visit = require("unist-util-visit")
+
+var tree = remark().parse("Some _emphasis_, **importance**, and `code`.")
+
+visit(tree, "text", visitor)
+
+function visitor(node) {
+  console.log(node)
+}
+```
+
+Here, it finds all text nodes and will `console.log` the nodes.
+
+Now with this, we can take the AST from our plugin and get working on adding functionality for headers:
+
+```js
+const visit = require("unist-util-visit")
+
+module.exports = ({ markdownAST }, pluginOptions) => {
+  visit(markdownAST, "heading", node => {
+    // Do stuff with heading nodes
+  })
+
+  return markdownAST
+}
+```
+
+Now we will visit all heading nodes and pass it into a transformer function which we can do whatever we wish.
```

Comment:
```suggestion
Now you will visit all heading nodes and pass it into a transformer function which you can do whatever we wish.
```

I can't come up with something better myself right now, but the "Now you will" should probably be replaced with some other introductions. You can probably remove it completely and just tell what to do, not what one will do now.

Reply:
Rephrased this paragraph. Hopefully it works better.